docs/rfcs/025-generation-numbers.md

```diff
@@ -0,0 +1,650 @@
+
+1. Deploy pageservers which understand the new index format and generation numbers
+   in keys, but do not write objects with generation numbers in the keys.
+2. Deploy pageservers that write objects with generation numbers in the keys.
+
+Old pageservers will be oblivious to generation numbers. That means that they can't
+read objects with generation numbers in the name.
+
+# Appendix A: Examples of use in high availability/failover
+
+The generation numbers proposed in this RFC are adaptable to a variety of different
+failover scenarios and models. The sections below sketch how they would work in practice.
+
+### Fast restart of a pageserver
+
+"fast" here means that the restart is done before any other element in the system
+has taken action in response to the node being down.
+
+- After restart, the node does no writes until it can obtain a fresh generation
+  number from the control plane.
+- Once issued a new generation number, the node synchronizes with control plane to learn of any
+  changes to its attachments, and receives a fast O(1) response informing it that it may continue
+  to operate with all its existing attachments (assuming the optimization from
+  "Optimizations" section for pageserver startup)
+- Node proceeds to operate as normal, trusting that all its local content
+  for attached tenants is up to date.
+
+The only difference between this and how the system behaves today is that
+the node generation number increases, and the pageserver's startup depends
+on availability of the control plane. The improvement is an increase in _safety_: we
+are no longer assuming that we may continue to write to all our attached
+tenants' data, we are confirming it before writing.
+
+### Unexpected "death" (possible network partition) of a pageserver
+
+Consider a network partition where somewhere there is a functioning node
+that believes it is the rightful holder of a particular node ID, and it
+can write to S3 even though it can't communicate with our control plane.
+
+This is the general case of a node dying: we must always assume that the
+unresponsive node might be network partitioned and still writing to S3.
+
+#### Case A: re-attachment to other nodes
+
+1. Let's say node 0 fails in a cluster of three nodes 0, 1, 2.
+2. Some external mechanism notices that the node is unavailable and initiates
+   movement of all tenants attached to that node to a different node. In
+   this example it would mean incrementing the attachment generation
+   of all tenants that were attached to node 0, and attaching them
+   to node 1 or 2 based on some distribution rule (this might be round
+   robin, or capacity based).
+3. A tenant which is now attached to node 1 will _also_ still be attached to node
+   0, from the perspective of node 0. Let's say it was originally attached with
+   generation suffix 00000001-0000-00000001 -- its new suffix on node 1
+   might be 00000002-0001-00000001 (attachment gen 2, on node 1).
```

Comment:
Would be helpful to repeat here what the three components of the suffix mean.

Reply:
done 8f7da22a